pilot/adapter/config/file/monitor.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"istio.io/istio/pilot/adapter/config/crd"
+	"istio.io/istio/pilot/model"
+)
+
+const (
+	defaultDuration = time.Second / 2
+)
+
+var (
+	supportedExtensions = map[string]bool{
+		".yaml": true,
+		".yml":  true,
+	}
+)
+
+// compareIds compares the IDs (i.e. Namespace, Type, and Name) of the two configs and returns
+// 0 if a == b, -1 if a < b, and 1 if a > b. Used for sorting config arrays.
+func compareIds(a, b *model.Config) int {
+	// Compare Namespace
+	if v := strings.Compare(a.Namespace, b.Namespace); v != 0 {
+		return v
+	}
+	// Compare Type
+	if v := strings.Compare(a.Type, b.Type); v != 0 {
+		return v
+	}
+	// Compare Name
+	return strings.Compare(a.Name, b.Name)
+}
+
+// ByKey is an array of config objects that is capable or sorting by Namespace, Type, and Name.
+type ByKey []*model.Config
```

Comment:
This feels like the wrong place for this to be public - if it's generally useful (I think it is) then we should probably stick it at `//pilot/model/byKey.go` (or even `sort.go`), otherwise keep it private.

Can be done in a follow-on PR.

Reply:
Oops, this wasn't supposed to be public.  An artifact of some intermediate hackery.  Fixed.